R/gh_token.R

```diff
@@ -164,21 +180,162 @@ gh_auth <- function(token) {
+  # credentials does: pat_url <- sprintf("https://%s@github.com", pat_user)
+
+  pat_url <- sprintf(
+    "%s://%s@%s",
+    prot_host[[1]], pat_user, prot_host[[2]]
+  )
+
+  if (isTRUE(force_new)) {
+    credentials::git_credential_forget(pat_url)
+  }
+
+  # credentials does the "hack" unconditionally, but I think this is better
+  if (is_interactive()) {
+    if(isTRUE(verbose)) {
+      message2("If prompted for GitHub credentials, enter your PAT in the password field")
+    }
+    askpass <- Sys.getenv('GIT_ASKPASS')
+    if (nchar(askpass)) {
+      # Hack to override prompt sentence to say "Token" instead of "Password"
+      withr::local_envvar(c(
+        GIT_ASKTOKEN = askpass,
+        GIT_ASKPASS = system.file(
+          "ask_token.sh", package = "gh", mustWork = TRUE
+        ),
+        GIT_ASKTOKEN_NAME = sprintf(
+          "Personal Access Token (PAT) for %s", host_url
+        )
+      ))
+    }
+  } else {
+    withr::local_envvar(c(
+      GIT_ASKPASS = system.file(
+        "empty_token.sh", package = "gh", mustWork = TRUE
+      )
+    ))
+  }
+
+  for(i in 1:3) {
+    cred <- credentials::git_credential_ask(pat_url, verbose = verbose)
+    if (length(cred$password) && !is.na(cred$password)) {
+      if (nchar(cred$password) < 40) {
+        message2("Please enter a token in the password field, not your master password! Let's try again :-)")
+        message2(sprintf(
+          "To generate a new token, visit: %s/settings/tokens",
+          host_url
+        ))
+        credentials::credential_reject(cred)
+        next
+      }
+      if (isTRUE(validate)) {
+        hx <- curl::handle_setheaders(curl::new_handle(), Authorization = paste("token", cred$password))
+        req <- curl::curl_fetch_memory(sprintf("%s/user", api_url), handle = hx)
+        if (req$status_code >= 400) {
+          message2("Authentication failed. Token invalid.")
+          credentials::credential_reject(cred)
+          next
+        }
+        if (isTRUE(verbose)) {
+          data <- jsonlite::fromJSON(rawToChar(req$content))
+          helper <- tryCatch(
+            credentials::credential_helper_get()[1],
+            error = function(e){"??"}
+          )
+          message2(sprintf(
+            "Using token stored for %s for user '%s' (credential helper: %s)",
+            host_url, data$login, helper
+          ))
+        }
+      }
+      return(
+        do.call(Sys.setenv, stats::setNames(list(cred$password), token_env_var))
+      )
+    }
+  }
+  if (isTRUE(verbose)) {
+    message2(sprintf(
+      "Failed to obtain a valid token for %s after 3 attempts",
+      host_url
+    ))
+  }
+  return(FALSE)
 }
 
-get_baseurl <- function(x) {
-  if (!any(grepl("^https?://", x))) stop("Only works with HTTP(S) protocols")
-  prot <- sub("^(https?://).*$", "\\1", x)
-  rest <- sub("^https?://(.*)$", "\\1", x)
-  host <- sub("/.*$", "", rest)
-  paste0(prot, host)
+# taken directly from credentials
+message2 <- function(...){
```

Comment:
I have not thought much about why credentials overrides `message()` (?). I just copied it over.

Reply:
This is probably because sys/git writes to the console directly, and if we don't flush then the R output and the sys/git output is mixed up.